File: LeetCode/greedy/CanPlaceFlowers.java
package greedy;

public class CanPlaceFlowers {
    public boolean canPlaceFlowers(int[] flowerbed, int n) {
        if (n == 0) {
            return true;
        }

        int length = flowerbed.length;

        for (int i = 0; i < length; i = i + 2) {

            if (flowerbed[i] == 0) {
                boolean right = (i + 1 == length) || flowerbed[i + 1] == 0;

                if (right) {
                    n--;
                    if (n == 0) {
                        return true;
                    }
                } else {
                    i++;
                }
            }
        }

        return n == 0;
    }

    public static void main(String[] args) {
        CanPlaceFlowers solution = new CanPlaceFlowers();
        int[] flowerbed1 = {1, 0, 0, 0, 1};
        System.out.println(solution.canPlaceFlowers(flowerbed1, 1)); // Output: true

        int[] flowerbed2 = {1, 0, 0, 0, 1};
        System.out.println(solution.canPlaceFlowers(flowerbed2, 2)); // Output: false

        int[] flowerbed3 = {1, 0, 1, 0, 1, 0, 1};
        System.out.println(solution.canPlaceFlowers(flowerbed3, 1)); // Output: false

        int[] flowerbed4 = {1, 0, 0, 0, 1, 0, 0};
        System.out.println(solution.canPlaceFlowers(flowerbed4, 2)); // Output: true
    }
}

File: LeetCode_Python/greedy/CanPlaceFlowers.py
class CanPlaceFlowers:
    def canPlaceFlowers(self, flowerbed: list[int], n: int) -> bool:
        count = 0
        i = 0

        while i < len(flowerbed):
            if (
                flowerbed[i] == 0
                and (i == 0 or flowerbed[i - 1] == 0)
                and (i == len(flowerbed) - 1 or flowerbed[i + 1] == 0)
            ):
                flowerbed[i] = 1
                count += 1
            i += 1

        return count >= n


# Time Complexity: O(n), where n is the length of the flowerbed
# Space Complexity: O(1), we are not using any extra space that grows with the input size

if __name__ == "__main__":
    solution = CanPlaceFlowers()

    print("Expected Output : True")
    print("Actual Output : ", solution.canPlaceFlowers([1, 0, 0, 0, 1], 1))

    print("Expected Output : False")
    print("Actual Output : ", solution.canPlaceFlowers([1, 0, 0, 0, 1], 2))
